### PR TITLE
Allow local users with an expiry date to be created

### LINKED ISSUE #71942: Local users with an expiry date cannot be created
##### SUMMARY
Local users with an expiry date cannot be created

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
ansible.builtin.user

##### ANSIBLE VERSION
```paste below
ansible 2.9.13
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.6/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 3.6.8 (default, Apr 16 2020, 01:36:27) [GCC 8.3.1 20191121 (Red Hat 8.3.1-5)]
```

##### OS / ENVIRONMENT

```paste below
cat /etc/os-release 
NAME="CentOS Linux"
VERSION="8 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="8"
PLATFORM_ID="platform:el8"
PRETTY_NAME="CentOS Linux 8 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:8"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-8"
CENTOS_MANTISBT_PROJECT_VERSION="8"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="8"
```

##### STEPS TO REPRODUCE

```yaml
---
- hosts: localhost
  tasks:
    - name: Create local user with expiry
      user:
        name: test_user
        expires: 1601093139
        local: yes
```

##### EXPECTED RESULTS

User `test_user` gets created locally with an expiry time of 1601093139

##### ACTUAL RESULTS

Playbook fails

```paste below
ansible-playbook 2.9.13
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.6/site-packages/ansible
  executable location = /usr/bin/ansible-playbook
  python version = 3.6.8 (default, Apr 16 2020, 01:36:27) [GCC 8.3.1 20191121 (Red Hat 8.3.1-5)]
Using /etc/ansible/ansible.cfg as config file
setting up inventory plugins
host_list declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
script declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
auto declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
Parsed /etc/ansible/hosts inventory source with ini plugin
Loading callback plugin default of type stdout, v2.0 from /usr/lib/python3.6/site-packages/ansible/plugins/callback/default.py

PLAYBOOK: user_failure.yml ****************************************************************************************************************************************************************************************
Positional arguments: user_failure.yml
verbosity: 4
connection: smart
timeout: 10
become_method: sudo
tags: ('all',)
inventory: ('/etc/ansible/hosts',)
forks: 5
1 plays in user_failure.yml

PLAY [localhost] **************************************************************************************************************************************************************************************************

TASK [Gathering Facts] ********************************************************************************************************************************************************************************************
task path: /home/ruediger/zw/user_failure.yml:2
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: ruediger
<127.0.0.1> EXEC /bin/sh -c 'echo ~root && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp `"&& mkdir "` echo /root/.ansible/tmp/ansible-tmp-1601043678.8671873-84732-75452603768298 `" && echo ansible-tmp-1601043678.8671873-84732-75452603768298="` echo /root/.ansible/tmp/ansible-tmp-1601043678.8671873-84732-75452603768298 `" ) && sleep 0'
Using module file /usr/lib/python3.6/site-packages/ansible/modules/system/setup.py
<127.0.0.1> PUT /root/.ansible/tmp/ansible-local-847254menfg02/tmptezho3pn TO /root/.ansible/tmp/ansible-tmp-1601043678.8671873-84732-75452603768298/AnsiballZ_setup.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1601043678.8671873-84732-75452603768298/ /root/.ansible/tmp/ansible-tmp-1601043678.8671873-84732-75452603768298/AnsiballZ_setup.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python3.6 /root/.ansible/tmp/ansible-tmp-1601043678.8671873-84732-75452603768298/AnsiballZ_setup.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /root/.ansible/tmp/ansible-tmp-1601043678.8671873-84732-75452603768298/ > /dev/null 2>&1 && sleep 0'
ok: [localhost]
META: ran handlers

TASK [Create local user with expiry] ******************************************************************************************************************************************************************************
task path: /home/ruediger/zw/user_failure.yml:4
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: ruediger
<127.0.0.1> EXEC /bin/sh -c 'echo ~root && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp `"&& mkdir "` echo /root/.ansible/tmp/ansible-tmp-1601043682.0082617-84816-139718076329706 `" && echo ansible-tmp-1601043682.0082617-84816-139718076329706="` echo /root/.ansible/tmp/ansible-tmp-1601043682.0082617-84816-139718076329706 `" ) && sleep 0'
Using module file /usr/lib/python3.6/site-packages/ansible/modules/system/user.py
<127.0.0.1> PUT /root/.ansible/tmp/ansible-local-847254menfg02/tmp967iqq6l TO /root/.ansible/tmp/ansible-tmp-1601043682.0082617-84816-139718076329706/AnsiballZ_user.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1601043682.0082617-84816-139718076329706/ /root/.ansible/tmp/ansible-tmp-1601043682.0082617-84816-139718076329706/AnsiballZ_user.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python3.6 /root/.ansible/tmp/ansible-tmp-1601043682.0082617-84816-139718076329706/AnsiballZ_user.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /root/.ansible/tmp/ansible-tmp-1601043682.0082617-84816-139718076329706/ > /dev/null 2>&1 && sleep 0'
fatal: [localhost]: FAILED! => {
    "changed": false,
    "invocation": {
        "module_args": {
            "append": false,
            "authorization": null,
            "comment": null,
            "create_home": true,
            "expires": 1601093139.0,
            "force": false,
            "generate_ssh_key": null,
            "group": null,
            "groups": null,
            "hidden": null,
            "home": null,
            "local": true,
            "login_class": null,
            "move_home": false,
            "name": "test_user",
            "non_unique": false,
            "password": null,
            "password_lock": null,
            "profile": null,
            "remove": false,
            "role": null,
            "seuser": null,
            "shell": null,
            "skeleton": null,
            "ssh_key_bits": 0,
            "ssh_key_comment": "ansible-generated on gauss",
            "ssh_key_file": null,
            "ssh_key_passphrase": null,
            "ssh_key_type": "rsa",
            "state": "present",
            "system": false,
            "uid": null,
            "update_password": "always"
        }
    },
    "msg": "Error parsing arguments: unknown option.\nUsage: luseradd [-irMn?] [-i|--interactive] [-r|--reserved]\n        [-c|--gecos=STRING] [-d|--directory=STRING] [-k|--skeleton=STRING]\n        [-s|--shell=STRING] [-u|--uid=NUM] [-g|--gid=STRING]\n        [-M|--nocreatehome] [-n|--nocreategroup] [-P|--plainpassword=STRING]\n        [-p|--password=STRING] [--commonname=STRING] [--givenname=STRING]\n        [--surname=STRING] [--roomnumber=STRING] [--telephonenumber=STRING]\n        [--homephone=STRING] [-?|--help] [--usage] [OPTION...] user\n",
    "name": "test_user",
    "rc": 1
}

PLAY RECAP ********************************************************************************************************************************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0   
```

This is caused by the fact that `luseradd` / `lusermod` at least on RedHat / Centos 6, 7, 8 and on Ubuntu 20 LTS do not support the `-e` option. I see the following options to fix this:

1. Just ignore `expires` in case `local` is `yes` or print a warning like the one if the user is not found in `/etc/passwd` for the local case when `expires` and `local` are set.
2. Fail on module level when `expires` and `local` are set.
3. If  `expires` is set and `local` is `yes` execute `lchage` after the user was created / modified to fix the expiry date.

Let me know if one of the above options would be acceptable (or another option) and I try to have a look for a patch / PR.

### PR DESCRIPTION
**Pull request by rpluem-vf**
The luseradd / lusermod commands do not support the -e option. Set
the expiry time in this case via lchage after the user was
created / modified.

##### SUMMARY

Fixes: #71942

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

user

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

**Comment by rpluem-vf** [comment]
ready_for_review

File: changelogs/fragments/fix_ansible_issue_71942.yaml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+bugfixes:
+  - >
+     user - Local users with an expiry date cannot be created as the luseradd /
+     lusermod commands do not support the -e option. Set the expiry time in
+     this case via lchage after the user was created / modified. (https://github.com/ansible/ansible/issues/71942)
```
Comment (samdoran) [inline changelogs/fragments/fix_ansible_issue_71942.yaml]:
```suggestion
     user - Local users with an expiry date cannot be created as the ``luseradd`` /
     ``lusermod`` commands do not support the ``-e`` option. Set the expiry time in
     this case via ``lchage`` after the user was created / modified. (https://github.com/ansible/ansible/issues/71942)
```

File: lib/ansible/modules/user.py [hunk 2]
```diff
@@ -674,7 +676,22 @@ def create_user_useradd(self):
...
-        if not self.local or rc != 0 or self.groups is None or len(self.groups) == 0:
+        if not self.local or rc != 0:
+            return (rc, err, out)
+
+        if self.expires is not None:
+            if self.expires < time.gmtime(0):
+                lexpires = -1
+            else:
+                # Convert seconds since Epoch to days since Epoch
+                lexpires = int(math.floor(float(self.module.params['expires']))) // 86400
```
Comment (samdoran) [inline lib/ansible/modules/user.py]: I'm not sure the `float` conversion is necessary. The `expires` parameter is specified as `type='float'` which means it has already gone through `check_type_float()` which means it has already been converted to a `float`.

I had to look into why `int(math.floor())` was necessary. `math.floor()` returns a `float` on Python 2 and an `int` in Python 3. I learned a bit from this line, so thanks for that!

File: lib/ansible/modules/user.py [hunk 3]
```diff
@@ -674,7 +676,22 @@ def create_user_useradd(self):
...
+            (rc, _err, _out) = self.execute_command([lchage_cmd, '-E', str(lexpires), self.name])
```
Comment (samdoran) [inline lib/ansible/modules/user.py]:
```suggestion
            (rc, _err, _out) = self.execute_command([lchage_cmd, '-E', to_native(lexpires), self.name])
```

File: lib/ansible/modules/user.py [hunk 4]
```diff
@@ -801,16 +820,23 @@ def modify_user_usermod(self):
...
                 # Convert days since Epoch to seconds since Epoch as struct_time
                 current_expire_date = time.gmtime(current_expires * 86400)
 
                 # Current expires is negative or we compare year, month, and day only
                 if current_expires < 0 or current_expire_date[:3] != self.expires[:3]:
-                    cmd.append('-e')
-                    cmd.append(time.strftime(self.DATE_FORMAT, self.expires))
+                    if self.local:
+                        # Convert seconds since Epoch to days since Epoch
+                        lexpires = int(math.floor(float(self.module.params['expires']))) // 86400
```
Comment (samdoran) [inline lib/ansible/modules/user.py]: Same here: I don't believe the inner `float()` conversion is necessary.

File: lib/ansible/modules/user.py [hunk 5]
```diff
@@ -830,7 +856,17 @@ def modify_user_usermod(self):
             cmd.append(self.name)
             (rc, err, out) = self.execute_command(cmd)
 
-        if not self.local or not (rc is None or rc == 0) or (len(lgroupmod_add) == 0 and len(lgroupmod_del) == 0):
+        if not self.local or not (rc is None or rc == 0):
+            return (rc, err, out)
+
+        if lexpires is not None:
+            (rc, _err, _out) = self.execute_command([lchage_cmd, '-E', str(lexpires), self.name])
```
Comment (samdoran) [inline lib/ansible/modules/user.py]:
```suggestion
            (rc, _err, _out) = self.execute_command([lchage_cmd, '-E', to_native(lexpires), self.name])
```

File: test/integration/targets/user/tasks/main.yml [hunk 6]
```diff
@@ -1135,3 +1135,336 @@
   when: ansible_facts.system in ['Linux']
   tags:
     - user_test_local_mode
+
```
Comment (samdoran) [inline test/integration/targets/user/tasks/main.yml:1138]: These tests are great. This `main.yml` file is a rather huge and I plan to break it up at some point. Could you move these tests into a separate tasks file and use `import_tasks` to make my future work a bit easier?

**Review by samdoran — CHANGES_REQUESTED** [review]
The change is good overall. Just a few minor changes are needed.

File: lib/ansible/modules/user.py (hunk 4 above)
Comment (rpluem-vf) [reply lib/ansible/modules/user.py]: I guess you are correct. Will change it and let us see what the testing says.

Comment (rpluem-vf) [reply lib/ansible/modules/user.py]: Done in https://github.com/ansible/ansible/pull/72022/commits/afd6e08b4fa3ef2e002f67dfc7e0efb6fbcc15f2

File: test/integration/targets/user/tasks/main.yml (hunk 6 above)
Comment (rpluem-vf) [reply test/integration/targets/user/tasks/main.yml:1138]: Done in https://github.com/ansible/ansible/pull/72022/commits/d5c473e34986277a0cd35390608e4cdca2a1eb5f

File: lib/ansible/modules/user.py (hunk 2 above)
Comment (rpluem-vf) [reply lib/ansible/modules/user.py]: I learned as well :-). I stumbled across this as the tests failed on some platforms before https://github.com/ansible/ansible/pull/72022/commits/1695b5ed8b33afee8fba34ae759eaaeb98abc15b and the difference was as you state that the failing platforms ran the tests on Python 2 while the succeeding ones ran on Python 3. Hence the great wealth of platforms tested was very helpful in detecting this issue.
Done in afd6e08.

**Review by rpluem-vf — COMMENTED** [review]
Hopefully things are fine now.

**Review by samdoran — COMMENTED** [review]
Thank you for addressing my comments so quickly.

**Comment by samdoran** [comment]
Please create a [backport PR](https://docs.ansible.com/ansible/latest/community/development_process.html#backporting-merged-prs) against `stable-2.9` and `stable-2.10`.

**Comment by rpluem-vf** [comment]
stable-2.10:  #72085
stable-2.9: #72086

**Pull request merged**